### PR TITLE
sampling : refactor + optimize penalties sampler

### PR DESCRIPTION
**Pull request by ggerganov**
Refactor, optimize and simplify the penalties sampler. It's position in `common_sampler` is now customizable, instead of being hardcoded at the front of the chain:

```bash
... --sampling-seq kep ...

sampler chain: logits -> logit-bias -> top-k -> penalties -> top-p -> dist
```

The main reason to allow this is that the penalties can be quite expensive to apply on full vocabulary. Now, they can be applied after a top-k sampler for example.

Also, the token count frequencies are now maintained in the sampler instead of being recreated on each token. Also, the `penalize_nl` option is removed since it is not relevant with new models.

### API Changes

- Change `llama_sampler_init_penalties()`

### Server API changes

- Remove `penalize_nl` parameter

### CODE REVIEW

**Comment by slaren** [comment]
> Also, the `ignore_eos` and `penalize_nl` options are removed since the former can be achieved through logit biases and the latter is not relevant with new models.

I don't think there is anything wrong with keeping `--ignore-eos` as a shortcut to the logit bias, although it is true that it is not as useful now as it was when it was added.

**Comment by ggerganov** [comment]
Restored the `--ignore-eos` options.

**Comment by slaren** [comment]
There is a pending issue from the initial refactor that setting a `--repeat-last-n` to -1 does not set it to `n_ctx`, unlike what the documentation says, if this cannot be fixed at least the documentation should be updated.
https://github.com/ggerganov/llama.cpp/blob/8faa1d4dd42f6cb26088ce7f5bbca5996b921685/common/arg.cpp#L886-L892

**Review by ngxson — APPROVED** [review]

**Comment by ggerganov** [comment]
Decided to move the penalties sampler at the end of the default sampling chain. This will change the default behaviour compared to `master`, but should result in significant sampling speed speedup (more than x5) when using any penalties.

File: common/common.cpp [hunk 1]
```diff
@@ -940,6 +940,21 @@ struct common_init_result common_init_from_params(common_params & params) {
         params.sampling.ignore_eos = false;
     }
 
+    if (params.sampling.ignore_eos) {
+        LOG_INF("%s: added EOS logit bias = %f\n", __func__, -INFINITY);
+        params.sampling.logit_bias.push_back({llama_token_eos(model), -INFINITY});
+    }
```
Comment (slaren) [inline common/common.cpp:946]: Noticed this doesn't really work for models with multiple eog tokens.
```suggestion
    if (params.sampling.ignore_eos) {
        for (llama_token i = 0; i < llama_n_vocab(model); i++) {
            if (llama_token_is_eog(model, i)) {
                LOG_INF("%s: added %s logit bias = %f\n", __func__, common_token_to_piece(lctx, i).c_str(), -INFINITY);
                params.sampling.logit_bias.push_back({i, -INFINITY});
            }
        }
    }
```

**Comment by ggerganov** [comment]
Ok, I'll move it back, but we have to figure out something else eventually. We cannot apply these penalties to the full vocabulary because the performance is significantly affected. Depending on your CPU and the number of parallel slots that you use, the slowdown from enabling either of these penalties can be really significant and I am pretty sure a lot of people out there are using it without even realizing the performance hit from this.

I don't use normally any repetition penalties and therefore haven't noticed this. But recently I started working on text-to-speech (TTS) models, and for some reason (that I don't yet understand completely), they seem to benefit from repetition penalties. And it does not matter if the rep penalty is applied before or after top-k. With the default sampling sequence on `master` these penalties will take a significant toll on the performance and it took me quite some time to realize this. So the defaults are not good and they should be updated somehow. The primary goal is to not affect the performance in a significant way and only allow it if the user really understands the implications. Simply setting a repetition penalty to 1.1 should not make any noticeable difference to the speed as it does now.

**Comment by MaggotHATE** [comment]
> So the defaults are not good and they should be updated somehow.

They are already not for LLMs too because `top_p` is not disabled by default, this has been discussed for a long time now. TTS models are new (I assume it's OuteTTS in this case), so it makes sense that they would require specific settings for optimal result - the same way as `top_p` is usually suggested to be turned off.

In general, it might be a good idea to have sampling parameters suggestions in some form, either in README or in console warnings. For TTS models it can be a warning to use repetition penalty, for LLMs - to switch `top_p` off. Performance-related warning can be added too.

**Comment by ggerganov** [comment]
By good defaults, I am focusing on the performance - not the generation quality.

I am worried about what is being applied before top-k. This is where the performance issues are because we haven't truncated the vocab yet. After top-k, you are free to put anything you like because it mainly depends on your use case (e.g. for FIM code completion, top-p makes sense IMO) and the performance will be good, as long as you didn't set a very big K.

**Comment by ggerganov** [comment]
Also, to clarify the significance of this issue even better, the performance penalty from enabling penalties stacks linearly with the number of parallel users of `llama-server`, because the sampling from the slot results is currently done sequentially. So the more users who enabled penalties, the slower the performance will be for all users, regardless if they are using penalties or not.

**Comment by MaggotHATE** [comment]
> a very big K

In that case, how much is "very big K"?

Theoretically, what is the K threshold value at which it's large enough to not affect creativity, but is smaller than vocab size enough to improve performance? If it exists, we can set the default value to this threshold and put `top_k` as the first sampler - but only if it doesn't change the result of `penalty repeat` significantly.

**Comment by ggerganov** [comment]
@ngxson I am trying to build the new WebUI and I get this error:

```bash
$ ▶ npm run build

> webui@0.0.0 build
> vite build

vite v5.4.11 building for production...
transforming (1) index.html
🌼   daisyUI 4.12.14
├─ ✔︎ 32 themes added		https://daisyui.com/docs/themes
╰─ ★ Star daisyUI on GitHub	https://github.com/saadeghi/daisyui

(node:87476) ExperimentalWarning: CommonJS module /llama.cpp/examples/server/webui/node_modules/tailwindcss/lib/lib/load-config.js is loading ES Module /llama.cpp/examples/server/webui/tailwind.config.js using require().
Support for loading ES Module in require() is an experimental feature and might change at any time
(Use `node --trace-warnings ...` to show where the warning was created)
file:///llama.cpp/examples/server/webui/tailwind.config.js:10
  plugins: [
           ^

ReferenceError: require is not defined
    at file:///llama.cpp/examples/server/webui/tailwind.config.js:10:12
    at ModuleJobSync.runSync (node:internal/modules/esm/module_job:395:35)
    at ModuleLoader.importSyncForRequire (node:internal/modules/esm/loader:329:47)
    at loadESMFromCJS (node:internal/modules/cjs/loader:1376:24)
    at Module._compile (node:internal/modules/cjs/loader:1528:5)
    at Object..js (node:internal/modules/cjs/loader:1698:10)
    at Module.load (node:internal/modules/cjs/loader:1303:32)
    at Function._load (node:internal/modules/cjs/loader:1117:12)
    at TracingChannel.traceSync (node:diagnostics_channel:322:14)
    at wrapModuleLoad (node:internal/modules/cjs/loader:218:24)

Node.js v23.3.0
```

Do you know how to fix it?

**Comment by ngxson** [comment]
It's a known nodejs 22.12 issue, I fixed it via https://github.com/ggerganov/llama.cpp/pull/10779

You just need to merge with latest master branch

File: examples/server/webui/src/main.js [hunk 2]
```diff
@@ -21,7 +21,7 @@ const CONFIG_DEFAULT = {
   systemMessage: 'You are a helpful assistant.',
   showTokensPerSecond: false,
   // make sure these default values are in sync with `common.h`
-  samplers: 'dkypmxt',
+  samplers: 'edkypmxt',
```
Comment (ggerganov) [inline examples/server/webui/src/main.js:24]: @ngxson Note that I made this change to the WebUI and AFAIU I have to build the new `index.html` with `npm run build`. But this command is failing on my computer with the error in the previous comment.

Comment (ngxson) [reply examples/server/webui/src/main.js:24]: Don't worry I can push the built index.html to this PR.

If you want to build it on your computer, maybe you can install a version manager like [nvm](https://github.com/nvm-sh/nvm), then `nvm use 22.11.0` to use the correct version

Comment (ngxson) [reply examples/server/webui/src/main.js:24]: note: we can also have a manually run CI job that build the frontend and output an artifact, I'll add this in the future

Comment (ggerganov) [reply examples/server/webui/src/main.js:24]: Ok thanks. I managed to install `22.11.0` with `nvm` and it works now 👍

**Comment by p-e-w** [comment]
Putting `top_k` first and defaulting it to a high value is a reasonable compromise. AFAIK, `top_k` is already special-cased because `top_k = 1` disables the entire sampler chain and disregards sampler order, even if it would have changed which token is in first position. `top_k` is a terrible sampler for steering generation anyway, and the value chosen by the user will usually be either `1` (if they want greedy sampling) or whatever the default is.

A `top_k` of `128` is already extremely generous. In practice, the 50th most probable token is often below `10^-7` already. IMO, llama.cpp could impose a maximum of `128` *even when the user chooses `0`,* that is, no limit. That would bring the performance benefits to users of frontends that always set a value for `top_k`, and typically default to `0`.

**Comment by MaggotHATE** [comment]
> AFAIK, `top_k` is already special-cased because `top_k = 1` disables the entire sampler chain and disregards sampler order

It was `temp`, and that feature was reworked [a while ago](https://github.com/ggerganov/llama.cpp/pull/9897), so technically speaking, all samplers are non-special now, especially with this PR - they all will be in one sampling chain.

As for `128` limit, the only times I saw a really large number of candidates were with `p_step` sampler - and it wasn't filtering out anything in those cases. So long as we don't have any `noise`-type sampler, a reasonable limit should not be a problem (if we keep customization for the user, of course).

**Comment by ggerganov** [comment]
I also feel like promoting the top-k sampler to the very front of the sampling chain would be good as it will cleanly solve the performance problems for default settings. And since the chain is fully customizable, we won't remove any functionality, but just make sure that it's not so easy to start penalizing the entire vocab.

Btw, in #9897 I completely forgot about the penalties sampler being in front of the top-k sampler. So the description there that `--top-k 1` is equivalent to greedy sampling is only true if there are no penalties enabled. Anyway, it's a small detail and if we move the top-k to the front, it wouldn't be an issue.

Will open a separate PR for that and we can discuss this further if necessary.

File: src/llama-sampling.cpp [hunk 3]
```diff
@@ -1421,76 +1417,40 @@ static void llama_sampler_penalties_accept(struct llama_sampler * smpl, llama_to
...
+    // if the ring buffer is full, remove the oldest token
+    if (ctx->prev.size() >= (size_t) ctx->penalty_last_n) {
+        const auto pop = ctx->prev.front();
+
+        ctx->token_count[pop]--;
+        if (ctx->token_count[pop] == 0) {
+            ctx->token_count.erase(pop);
+        }
+    }
+
```
Comment (ggerganov) [inline src/llama-sampling.cpp:1431]: This change needs extra attention for correctness.

**Pull request merged**